alf/utils/tensor_utils.py

```diff
@@ -80,3 +81,105 @@ def to_tensor(data, dtype=None):
         # as_tensor reuses the underlying data store of numpy array if possible.
         data = torch.as_tensor(data, dtype=dtype).detach()
     return data
+
+
+def global_norm(tensors):
+    """Adapted from TF's version.
+    Computes the global norm of a nest of tensors. Given a nest of tensors
+    `tensors`, this function returns the global norm of all tensors in `tensors`.
+    The global norm is computed as:
+
+        `global_norm = sqrt(sum([l2norm(t)**2 for t in t_list]))`
+
+    Any entries in `tensors` that are of type None are ignored.
+
+    Args:
+        tensors (nested Tensor): a nest of tensors
+
+    Returns:
+        norm (Tensor): a scalar tensor
+    """
+    assert alf.nest.is_nested(tensors), "tensors must be a nest!"
+    tensors = alf.nest.flatten(tensors)
+    return torch.sqrt(
+        sum([
+            math_ops.square(torch.norm(torch.reshape(t, [-1])))
+            for t in tensors if t is not None
+        ]))
+
+
+def clip_by_global_norm(tensors, clip_norm, use_norm=None):
+    """Adapted from TF's version.
+    Clips values of multiple tensors by the ratio of `clip_norm` to the global
+    norm.
+
+    Given a nest of tensors `tensors`, and a clipping norm threshold `clip_norm`,
+    this function clips the tensors *in place* and returns the global norm
+    (`global_norm`) of all tensors in `tensors`. Optionally, if you've already
+    computed the global norm for `tensors`, you can specify the global norm with
+    `use_norm`.
+
+    To perform the clipping, each `tensor` are set to:
+        tensor * clip_norm / max(global_norm, clip_norm)
+    where:
+        global_norm = sqrt(sum([l2norm(t)**2 for t in tensors]))
+
+    If `clip_norm > global_norm` then the entries in `tensors` remain as they are,
+    otherwise they're all shrunk by the global ratio.
+
+    Any of the entries of `tensors` that are of type `None` are ignored.
+
+    Args:
+        tensors (nested Tensor): a nest of tensors to be clipped
+        clip_norm (float or Tensor): a positive floating scalar
+        use_norm (float or Tensor): the global norm to use. If None,
+            `global_norm()` will be used to compute the norm.
+    Returns:
+        global_norm (Tensor): a scalar tensor representing the global norm. If
+            `use_norm` is provided, it will be returned instead.
+    """
+    assert alf.nest.is_nested(tensors), "tensors must be a nest!"
```

Comment:
why does it have to be nested?

Reply:
Because if it's just a single tensor, then the name "global" seems confusing. If TF's version, it has to be a list.